app/models/miq_smart_proxy_worker/runner.rb

```diff
@@ -54,18 +50,24 @@ def heartbeat_thread
     end
   end
 
-  def do_work
-    if @tid.nil? || !@tid.alive?
-      unless @tid.try(:status)
-        dead_tid, @tid = @tid, nil
-        _log.info("#{log_prefix} Waiting for the Heartbeat Thread to exit...")
-        dead_tid.join # raise the exception the dead thread failed with
-      end
+  def ensure_heartbeat_thread_started
+    # start the thread and return if it has not been started yet
+    if @tid.nil?
+      @tid = start_heartbeat_thread
+      return
+    end
+
+    # if the thread is dead, start a new one and kill it if it is aborting
+    if !@tid.alive?
+      @tid.kill if @tid.status == "aborting"
```

Comment:
It's amazing how much easier this is to grok by splitting apart that conditional into two conditional checks.